### PR TITLE
Add T.<unchecked_let>

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Needed for some Stripe-internal use cases.

### Test plan

See included automated tests.

### CODE REVIEW

File: cfg/builder/builder_walk.cc [hunk 1]
```diff
@@ -675,7 +675,9 @@ BasicBlock *CFGBuilder::walk(CFGContext cctx, ast::Expression *what, BasicBlock
             [&](ast::Cast *c) {
                 core::LocalVariable tmp = cctx.newTemporary(core::Names::castTemp());
                 current = walk(cctx.withTarget(tmp), c->arg.get(), current);
-                current->exprs.emplace_back(cctx.target, c->loc, make_unique<Cast>(tmp, c->type, c->cast));
+                if (c->cast != core::Names::uncheckedLet()) {
+                    current->exprs.emplace_back(cctx.target, c->loc, make_unique<Cast>(tmp, c->type, c->cast));
```
Comment (DarkDimius) [inline cfg/builder/builder_walk.cc:679]: will fail in `strong`, but unsure if it failed before

File: cfg/builder/builder_walk.cc [hunk 2]
```diff
@@ -675,7 +675,9 @@ BasicBlock *CFGBuilder::walk(CFGContext cctx, ast::Expression *what, BasicBlock
             [&](ast::Cast *c) {
                 core::LocalVariable tmp = cctx.newTemporary(core::Names::castTemp());
                 current = walk(cctx.withTarget(tmp), c->arg.get(), current);
-                current->exprs.emplace_back(cctx.target, c->loc, make_unique<Cast>(tmp, c->type, c->cast));
+                if (c->cast != core::Names::uncheckedLet()) {
```
Comment (elliottt) [inline cfg/builder/builder_walk.cc:678]: Will anything fail if `cctx.target` doesn't get populated? Could you add cfg exp tests for some of the `T::Enum` tests and make sure that everything makes sense?

**Comment by jez** [comment]
> Could you add CFG exp files for some of the T::Enum tests so that we can validate that not setting cctx.target for uncheckedLet doesn't cause problems?

There is a snapshot for T::Enum, and it changes in this PR. What are you asking for?

File: test/testdata/rewriter/t_enum_snapshot.rb.cfg.exp [hunk 3]
```diff
@@ -46,7 +46,7 @@ subgraph "cluster_::<Class:MyEnum>#<static-init>" {
...
...<cfgAlias>$76: T.class_of(MyEnum::Z))\l<cfgAlias>$79: T.class_of(Sorbet::Private::Static) = alias <C Static>\l<cfgAlias>$81: T.class_of(MyEnum::Z) = alias <C Z$1>\l<statTemp>$77: Sorbet::Private::Static::Void = <cfgAlias>$79: T.class_of(Sorbet::Private::Static).keep_for_ide(<cfgAlias>$81: T.class_of(MyEnum::Z))\l<cfgAlias>$84: T.class_of(Sorbet::Private::Static) = alias <C Static>\l<cfgAlias>$86: T.class_of(MyEnum) = alias <C MyEnum>\l<statTemp>$82: Sorbet::Private::Static::Void = <cfgAlias>$84: T.class_of(Sorbet::Private::Static).keep_for_ide(<cfgAlias>$86: T.class_of(MyEnum))\l<cfgAlias>$91: T.class_of(Sorbet::Private::Static) = alias <C Static>\l<cfgAlias>$93: T.class_of(MyEnum::Z) = alias <C Z$1>\l<statTemp>$89: Sorbet::Private::Static::Void = <cfgAlias>$91: T.class_of(Sorbet::Private::Static).keep_for_typechecking(<cfgAlias>$93: T.class_of(MyEnum::Z))\l<cfgAlias>$96: T.class_of(MyEnum::Z) = alias <C Z$1>\l<castTemp>$94: MyEnum::Z = <cfgAlias>$96: T.class_of(MyEnum::Z).new(<self>: T.class_of(MyEnum))\l<C Z>$88: MyEnum::Z = <castTemp>$94\l<blockReturnTemp>$17: NilClass = nil\l<blockReturnTemp>$98: T.noreturn = blockreturn<enums> <blockReturnTemp>$17: NilClass\l<unconditional>\l"
```
Comment (jez) [inline test/testdata/rewriter/t_enum_snapshot.rb.cfg.exp:49]: this diff looks ugly so here's a slightly better `git diff --word-diff`

image

**Pull request merged**